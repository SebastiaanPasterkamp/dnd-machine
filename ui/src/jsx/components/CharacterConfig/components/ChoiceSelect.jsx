import React from 'react';
import PropTypes from 'prop-types';
import {
    entries,
    filter,
    findIndex,
    flow,
    map,
} from 'lodash/fp';

import TabComponent from '../../TabComponent';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';

import CharacterConfig from '../CharacterConfig';
import MatchesFilters from '../utils/MatchesFilters';

export class ChoiceSelect extends React.Component
{
    constructor(props) {
        super(props);
        this.onTabChange = this.onTabChange.bind(this);
    }

    getTabs() {
        const { options, filter: filters } = this.props;
        return filter(
            (option) => !option.hidden && MatchesFilters(option, filters)
        )(options);
    }

    componentDidMount() {
        const { selected } = this.props;
        if (selected === null || selected === undefined) {
            this.onTabChange(0);
        }
    }

    onTabChange(index) {
        const { uuid, setState, selected } = this.props;
        const tabs = this.getTabs();
        if (
            index >= 0
            && index < tabs.length
            && tabs[index].uuid !== selected
        ) {
            setState({
                selected: tabs[index].uuid,
            });
        }
    }

    render() {
        const { selected } = this.props;
        const tabs = this.getTabs();

        if (!tabs.length) {
            return null;
        }

        const activeTab = findIndex({ uuid: selected }, tabs) || 0;

        return (
            <TabComponent
                tabConfig={tabs}
                activeTab={activeTab >= 0 ? activeTab : 0}
                onTabChange={this.onTabChange}
            >
                {flow(entries, map(
                    ([index, tab]) => (
                        <CharacterConfig
                            key={ index }
                            config={ [tab] }
                        />
                    )
                ))(tabs)}
            </TabComponent>
        );
    }
};

ChoiceSelect.propTypes = {
    type: PropTypes.oneOf(['choice']).isRequired,
    uuid: PropTypes.string.isRequired,
    setState: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(
        PropTypes.shape({
            description: PropTypes.string,
        })
    ).isRequired,
    selected: PropTypes.string,
};

ChoiceSelect.defaultProps = {
    type: 'choice',
    selected: null,
};

export default CharacterEditorWrapper(ChoiceSelect);
