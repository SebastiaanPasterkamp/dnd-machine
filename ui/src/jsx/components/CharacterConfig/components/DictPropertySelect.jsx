import React from 'react';
import PropTypes from 'prop-types';
import { sprintf } from 'sprintf-js';
import MDReactComponent from 'markdown-react-js';
import {
    keys,
    reduce,
} from 'lodash/fp';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';

export const DictPropertySelect = function({ hidden, current })
{
    if (hidden || !current) {
        return null;
    }

    let value = current.description;
    const defaulted = reduce(
        (defaulted, key) => {
            if (!key.match(/_default$/)) {
                return defaulted;
            }
            const field = key.replace(/_default$/, '');
            return {
                [field]: current[key],
                ...defaulted,
            };
        },
        {}
    )(keys(current));
    try {
        value = sprintf(
            current.description || '',
            { ...defaulted, ...current }
        );
    } catch(err) {
        console.error(err, current);
    }

    return (
        <MDReactComponent text={value} />
    );
};

DictPropertySelect.propTypes = {
    type: PropTypes.oneOf(['dict']).isRequired,
    uuid: PropTypes.string.isRequired,
    current: PropTypes.object,
    dict: PropTypes.object,
    hidden: PropTypes.bool,
};

export default CharacterEditorWrapper(DictPropertySelect);
