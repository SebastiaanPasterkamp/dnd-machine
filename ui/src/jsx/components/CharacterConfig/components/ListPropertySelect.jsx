import React from 'react';
import PropTypes from 'prop-types';
import {
    concat,
    countBy,
    entries,
    find,
    filter,
    flow,
    forEach,
    includes,
    map,
    uniq,
    without,
} from 'lodash/fp';

import {
    BaseTagContainer,
    Tag,
    TagBadge,
} from '../../BaseTagContainer';
import LazyComponent from '../../LazyComponent';
import SingleSelect from '../../SingleSelect';

import { memoize } from '../../../utils';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';
import ListsToItemsWrapper from '../../../hocs/ListsToItemsWrapper';
import MatchesFilters from '../utils/MatchesFilters';

export class ListPropertySelect extends LazyComponent
{
    constructor(props) {
        super(props);
        this.state = {
            showSelect: false,
            disabled: true,
        };
        this.memoize = memoize.bind(this);
        this.onAdd = this.onAdd.bind(this);
        this.onDelete = this.onDelete.bind(this);
        this.onChange = this.onChange.bind(this);
    }

    getId({ id, code, name }) {
        if (id !== undefined) {
            return id;
        }
        if (code !== undefined) {
            return code;
        }
        return name;
    }

    getLabel({ label, name }) {
        if (label !== undefined) {
            return label;
        }
        return name;
    }

    static getDerivedStateFromProps(props, state) {
        const { added, removed, given, add, limit, replace, current } = props;

        const showSelect = Boolean(
            (added.length - removed.length) < add
            || current.length < limit
        );
        const disabled = removed.length >= replace;

        if (showSelect !== state.showSelect
            || disabled !== state.disabled
        ) {
            return { showSelect, disabled };
        }
        return null;
    }

    componentDidMount() {
        const { given } = this.props;
        if (given.length) {
            this.onChange();
        }
    }

    onChange(update) {
        const {
            added, removed, given, onChange,
        } = { ...this.props, ...update };
        onChange({
            added: concat(added, given),
            removed,
        });
    }

    onDelete(id) {
        return this.memoize(
            `delete-${id}`,
            () => {
                const {
                    added, removed, replace, multiple, setState,
                } = this.props;
                const state = { added, removed };
                if (includes(id, added)) {
                    state.added = without([ id ], added);
                } else if (removed.length < replace) {
                    state.removed = concat(removed, [id]);
                    if (!multiple) {
                        state.removed = uniq(state.removed);
                    }
                }
                setState(state);
                this.onChange(state);
            }
        );
    }

    onAdd(id) {
        const {
            added, removed, add, limit, current, multiple, setState,
        } = this.props;

        if (limit) {
            if (current.length >= limit) {
                return false;
            }
        } else if (added.length >= (add + removed.length)) {
            return false;
        }

        const state = { added, removed };
        if (includes(id, removed)) {
            state.removed = without([ id ], removed);
        } else {
            state.added = concat(added, [id]);
            if (!multiple) {
                state.added = uniq(state.added);
            }
        }

        setState(state);
        this.onChange(state);
    }

    findItem(id, _default=undefined) {
        _default = _default || { label: `${id}`, color: 'bad' };
        const { items } = this.props;
        const match = find({id: id}, items)
                    || find({code: id}, items)
                    || find({name: id}, items);

        if (!match && _default !== undefined) {
            return _default;
        }

        const item = match || {
            ..._default,
            id,
        };

        return ({
            id: this.getId(item),
            label: this.getLabel(item),
            description: item.description,
        });
    }

    getValue() {
        const { added, removed, given, current, replace } = this.props;
        const { disabled } = this.state;

        const tags = concat(
            map(
                id => ({
                    ...this.findItem(id),
                    color: 'info',
                })
            )(added),
            map(
                id => ({
                    ...this.findItem(id),
                    color: 'good',
                    disabled: true,
                })
            )(given)
        );

        let upgrade = countBy('id', tags);
        let downgrade = countBy('id', removed);
        forEach(
            (count, id) => {
                upgrade[id] = (upgrade[id] || 0) - count;
            }
        )(downgrade);

        return concat(
            tags,
            filter(null, map(
                id => {
                    if (includes(id, given)) {
                        return null;
                    }
                    if (includes(id, removed)) {
                        return null;
                    }
                    if ((upgrade[id] || 0) > 0) {
                        upgrade[id] -= 1;
                        return null;
                    }
                    const _current = this.findItem(id, null);
                    if (!_current) {
                        return null;
                    }
                    return ({
                        ..._current,
                        color: disabled ? null : 'warning',
                        disabled,
                    });
                }
            )(current))
        );
    }

    renderSelect() {
        const {
            multiple,
            filter: filters,
            items,
            current,
            added,
            removed,
            given,
            limit,
            add,
            replace,
        } = this.props;
        const { showSelect } = this.state;

        if (!items.length) return null;

        if (!showSelect) {
            return null;
        }

        const values = [...given, ...current, ...added];
        const filtered = filter(
            item => (
                multiple
                || includes( this.getId(item), removed )
                || !includes( this.getId(item), values )
            ),
            filter(
                item => (
                    includes( this.getId(item), removed )
                    || MatchesFilters(item, filters)
                ),
                items
            )
        );

        if (!filtered.length) return null;

        return (
            <SingleSelect
                emptyLabel="Add..."
                items={filtered}
                setState={this.onAdd}
            />
        );
    }

    render() {
        const { hidden } = this.props;
        if (hidden) {
            return null;
        }

        return (
            <BaseTagContainer>
                {this.renderSelect()}

                {flow(entries, map(([i, tag]) => (
                    <Tag
                        key={`tag-${i}`}
                        onDelete={this.onDelete(tag.id)}
                        {...tag}
                    />
                )))(this.getValue())}
            </BaseTagContainer>
        );
    }

};

ListPropertySelect.propTypes = {
    type: PropTypes.oneOf(['list']).isRequired,
    uuid: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    setState: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.object),
    given: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ])
    ),
    current: PropTypes.array,
    limit: PropTypes.number,
    added: PropTypes.array,
    removed: PropTypes.array,
    add: PropTypes.number,
    replace: PropTypes.number,
    filter: PropTypes.object,
    multiple: PropTypes.bool,
    hidden: PropTypes.bool,
};

ListPropertySelect.defaultProps = {
    added: [],
    removed: [],
    given: [],
    current: [],
    limit: 0,
    add: 0,
    replace: 0,
    multiple: false,
    filter: {},
    items: [],
};

export default ListsToItemsWrapper(
    CharacterEditorWrapper(
        ListPropertySelect,
        {
            current: true,
        }
    ),
    'items'
);
