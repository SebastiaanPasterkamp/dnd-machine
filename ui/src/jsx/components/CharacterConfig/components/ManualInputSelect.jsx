import React from 'react';
import PropTypes from 'prop-types';

import LazyComponent from '../../LazyComponent';
import InputField from '../../InputField';
import MarkdownTextField from '../../MarkdownTextField';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';

export const ManualInputSelect = function({
    onChange, setState, current, placeholder, markup,
}) {
    this.onSetState = this.onSetState || function(current) {
        setState({ current });
        onChange(current);
    }.bind(this);

    if(markup) {
        return (
            <MarkdownTextField
                placeholder={placeholder}
                value={current}
                rows={5}
                setState={this.onSetState}
            />
        );
    }

    return (
        <InputField
            placeholder={placeholder}
            value={current}
            setState={this.onSetState}
        />
    );
};

ManualInputSelect.propTypes = {
    type: PropTypes.oneOf(['manual']).isRequired,
    uuid: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    setState: PropTypes.func.isRequired,
    current: PropTypes.string,
    placeholder: PropTypes.string,
    markup: PropTypes.bool,
};

ManualInputSelect.defaultProps = {
    current: '',
    placeholder: '',
    markup: false,
};

export default CharacterEditorWrapper(
    ManualInputSelect,
    {
        current: true,
    }
);
