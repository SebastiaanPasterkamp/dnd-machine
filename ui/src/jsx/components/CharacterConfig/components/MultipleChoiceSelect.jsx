import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import TagContainer from '../../TagContainer';

import CharacterConfig from '../CharacterConfig';
import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';
import MatchesFilters from '../utils/MatchesFilters';

export class MultipleChoiceSelect extends React.Component
{
    constructor(props) {
        super(props);
        this.state = {
            showSelect: false,
            disabled: true,
        };
        this.onAdd = this.onAdd.bind(this);
        this.onDelete = this.onDelete.bind(this);
    }

    static getDerivedStateFromProps(props, state) {
        const { added, removed, add, limit, replace } = props;
        const current = MultipleChoiceSelect.getCurrent(props);

        const showSelect = Boolean(
            (added.length - removed.length) < add
            || current.length < limit
        );
        const disabled = removed.length >= replace;

        if (showSelect !== state.showSelect
            || disabled !== state.disabled
        ) {
            return { showSelect, disabled };
        }
        return null;
    }

    onAdd(uuid) {
        const {
            added, removed, filtered, limit, add, replace, setState,
        } = this.props;
        const current = MultipleChoiceSelect.getCurrent(this.props);

        if (limit) {
            if (current.length >= limit) {
                return false;
            }
        } else if (added.length >= (add + removed.length)) {
            return false;
        }

        const state = { added, removed, filtered };
        if (_.includes(removed, uuid)) {
            state.removed = _.without(removed, uuid);
        } else if (added.length < (add + removed.length)) {
            state.filtered = _.without(filtered, uuid);
            state.added = _.concat(added, [uuid]);
        }
        setState(state);
    }

    onDelete(uuid) {
        const {
            added, removed, filtered, limit, add, replace, setState,
        } = this.props;
        const state = { added, removed, filtered };

        if (_.includes(added, uuid)) {
            state.added = _.without(added, uuid);
            state.filtered = _.concat(removed, [uuid]);
        } else if (removed.length < replace) {
            state.removed = _.concat(removed, [uuid]);
        }
        setState(state);
    }

    onSetState() {}

    static getCurrent(props, state) {
        const { added, removed, filtered, options, getCurrent } = props;

        const current = _.chain(options)
            .filter(option => {
                if (_.includes(added, option.uuid)) {
                    return true;
                }
                if (_.includes(filtered, option.uuid)) {
                    return false;
                }
                if (_.includes(removed, option.uuid)) {
                    return false;
                }
                const path = (
                    _.get(option, 'path')
                    || _.get(option, ['config', 0, 'path'])
                );
                return getCurrent(path);
            })
            .map(option => option.uuid)
            .value();
        return current;
    }

    render() {
        const {
            added, removed, filtered, options, limit, add, replace,
            filter: filters,
        } = this.props;
        const { showSelect, disabled } = this.state;

        const current = MultipleChoiceSelect.getCurrent(this.props);
        const itemsDisabled = disabled || removed.length >= replace;
        const items = _.chain(options)
            .filter(option => !option.hidden && MatchesFilters(option, filters))
            .map(option => {
                const isNew = _.includes(added, option.uuid);
                return {
                    id: option.uuid,
                    label: option.label,
                    color: isNew ? 'info' : 'warning',
                    disabled: !isNew && itemsDisabled,
                };
            })
            .value();

        return (
            <div>
                <TagContainer
                    value={current}
                    items={items}
                    onAdd={this.onAdd}
                    onDelete={this.onDelete}
                    setState={this.onSetState}
                    showSelect={showSelect}
                />
                {_.map(
                    options,
                    (config, i) => _.includes(current, config.uuid) ? (
                        <CharacterConfig
                            key={i}
                            config={ [config] }
                        />
                    ) : null
                )}
            </div>
        );
    }
};

MultipleChoiceSelect.propTypes = {
    type: PropTypes.oneOf(['multichoice']).isRequired,
    uuid: PropTypes.string.isRequired,
    setState: PropTypes.func.isRequired,
    getCurrent: PropTypes.func.isRequired,
    options: PropTypes.arrayOf(PropTypes.object),
    description: PropTypes.string,
    include: PropTypes.number,
    added: PropTypes.array,
    removed: PropTypes.array,
    filtered: PropTypes.array,
    limit: PropTypes.number,
    add: PropTypes.number,
    replace: PropTypes.number,
};

MultipleChoiceSelect.defaultProps = {
    added: [],
    removed: [],
    filtered: [],
    options: [],
    description: '',
    limit: 0,
    add: 0,
    replace: 0,
};

export default CharacterEditorWrapper(
    MultipleChoiceSelect,
    {
        character: true,
    }
);
