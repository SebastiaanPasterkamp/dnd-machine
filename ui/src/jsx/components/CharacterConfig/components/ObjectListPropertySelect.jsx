import React from 'react';
import PropTypes from 'prop-types';
import {
    concat,
    countBy,
    find,
    findIndex,
    filter,
    forEach,
    includes,
    isPlainObject,
    map,
    omitBy,
    sumBy,
    uniq,
    without,
} from 'lodash/fp';

import {
    BaseTagContainer,
    Tag,
    TagBadge,
} from '../../BaseTagContainer';
import SingleSelect from '../../SingleSelect';

import { memoize } from '../../../utils';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';
import ListsToItemsWrapper from '../../../hocs/ListsToItemsWrapper';
import MatchesFilters from '../utils/MatchesFilters';

export class ObjectListPropertySelect extends React.Component
{
    constructor(props) {
        super(props);
        this.state = {
            showSelect: false,
            disabled: true,
        };
        this.memoize = memoize.bind(this);
        this.onAdd = this.onAdd.bind(this);
        this.onDelete = this.onDelete.bind(this);
    }

    static getDerivedStateFromProps(props, state) {
        const { added, removed, given, current, add, limit, replace } = props;
        const cntAdded = sumBy(a => a.count || 1, added);
        const cntCurrent = sumBy(c => c.count || 1, current);
        const cntRemoved = sumBy(r => r.count || 1, removed);

        const showSelect = (cntAdded - cntRemoved) < add || cntCurrent < limit;
        const disabled = cntRemoved >= replace;

        if (showSelect !== state.showSelect || disabled !== state.disabled) {
            return { showSelect, disabled };
        }
        return null;
    }

    getNeedle(item) {
        const { type, id } = item || {};
        return omitBy(
            (value) => value === undefined
        )({type, id});
    }

    increase(item, set) {
        const { multiple } = this.props;
        const needle = this.getNeedle(item);

        const idx = findIndex(needle, set);
        if (idx >= 0) {
            return [
                ...set.slice(0, idx),
                {...set[idx], ...item, count: multiple ? set[idx].count + 1 : undefined},
                ...set.slice(idx + 1),
            ];
        }

        return [...set, {...item, count: multiple ? 1 : undefined}];
    }

    decrease(item, set) {
        const { multiple } = this.props;
        const needle = this.getNeedle(item);

        const idx = findIndex(needle, set);
        if (multiple && set[idx].count > 1) {
            return [
                ...set.slice(0, idx),
                {...set[idx], ...item, count: set[idx].count - 1},
                ...set.slice(idx + 1),
            ];
        }

        return [
            ...set.slice(0, idx),
            ...set.slice(idx + 1),
        ];
    }

    onDelete(item) {
        const { type, id } = item;
        const key = `delete-${type}-${id}`;
        const needle = this.getNeedle(item);

        return this.memoize(key, () => {
            const { added, removed, replace, setState } = this.props;
            const cntRemoved = sumBy(r => r.count || 1, removed);

            const state = { added, removed };
            if (find(needle, added)) {
                state.added = this.decrease(item, added);
            } else if (cntRemoved < replace) {
                state.removed = this.increase(item, removed);
            }

            setState(state);
        });
    }

    onAdd(item) {
        const { items, added, removed, setState } = this.props;
        const needle = this.getNeedle(item);

        const state = { added, removed };
        if (find(needle, removed)) {
            state.removed = this.decrease(item, removed);
        } else {
            state.added = this.increase(item, added);
        }

        setState(state);
    }

    getValue() {
        const {
            added, removed, items, current, given, replace, uuid
        } = this.props;
        const { disabled } = this.state;

        return map(
            (item) => {
                const needle = this.getNeedle(item);

                const isAdded = find(needle, added) !== undefined;
                const isGiven = find(needle, given) !== undefined;
                const isRemoved = find(needle, removed) !== undefined;

                return {
                    ...item,
                    color: isAdded ? 'info' : (
                        isRemoved ? 'bad' : (
                            isGiven ? 'good' : (
                                disabled && !isAdded ? null : 'warning'
                            )
                        )
                    ),
                    disabled: (disabled || isGiven) && !isAdded,
                };
            }
        )(current);
    }

    renderSelect() {
        const {
            multiple, items, current, removed,
            filter: filters,
        } = this.props;
        const { showSelect } = this.state;

        if (!items.length) return null;

        if (!showSelect) {
            return null;
        }

        const filtered = filter(
            item => {
                const needle = this.getNeedle(item);
                if (find(needle, removed)) {
                    return true;
                }
                if (!MatchesFilters(item, filters)) {
                    return false;
                }
                if (multiple) {
                    return true;
                }
                if (find(needle, current)) {
                    return false;
                }
                return true;
            }
        )(items);

        if (!filtered.length) return null;

        return (
            <SingleSelect
                emptyLabel="Add..."
                items={filtered}
                setState={this.onAdd}
                objects={true}
                filterable={true}
            />
        );
    }

    render() {
        const { hidden } = this.props;

        if (hidden) {
            return null;
        }

        return (
            <BaseTagContainer>
                {this.renderSelect()}

                {map(
                    tag => (
                        <Tag
                            key={`tag-${tag.type}-${tag.id}`}
                            onDelete={this.onDelete(tag)}
                            {...tag}
                        >
                            {tag.count > 1 ? (
                                <TagBadge>
                                    &times; {tag.count}
                                </TagBadge>
                            ) : null}
                        </Tag>
                    )
                )(this.getValue())}
            </BaseTagContainer>
        );
    }
};

const itemType = {
    id: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
    ]).isRequired,
    type: PropTypes.string,
    name: PropTypes.string,
    description: PropTypes.string,
    count: PropTypes.number,
};

ObjectListPropertySelect.propTypes = {
    type: PropTypes.oneOf(['objectlist']).isRequired,
    uuid: PropTypes.string.isRequired,
    setState: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]),
        name: PropTypes.string,
        description: PropTypes.string,
    })),
    given: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.shape(itemType),
            PropTypes.string,
            PropTypes.number,
        ]),
    ),
    current: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.shape(itemType),
            PropTypes.string,
            PropTypes.number,
        ]),
    ),
    limit: PropTypes.number,
    added: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.shape(itemType),
            PropTypes.string,
            PropTypes.number,
        ]),
    ),
    removed: PropTypes.arrayOf(
        PropTypes.oneOfType([
            PropTypes.shape(itemType),
            PropTypes.string,
            PropTypes.number,
        ]),
    ),
    add: PropTypes.number,
    replace: PropTypes.number,
    filter: PropTypes.arrayOf(
        PropTypes.object
    ),
    multiple: PropTypes.bool,
    hidden: PropTypes.bool,
};

ObjectListPropertySelect.defaultProps = {
    added: [],
    removed: [],
    given: [],
    current: [],
    limit: 0,
    add: 0,
    replace: 0,
    multiple: false,
    filter: [],
    items: [],
    hidden: false,
};

export default ListsToItemsWrapper(
    CharacterEditorWrapper(ObjectListPropertySelect),
    'items'
);
