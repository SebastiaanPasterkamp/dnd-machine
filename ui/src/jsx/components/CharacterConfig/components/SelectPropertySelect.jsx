import React from 'react';
import PropTypes from 'prop-types';

import LazyComponent from '../../LazyComponent';
import SingleSelect from '../../SingleSelect';

import CharacterEditorWrapper from '../hocs/CharacterEditorWrapper';
import ListsToItemsWrapper from '../../../hocs/ListsToItemsWrapper';

export class SelectPropertySelect extends LazyComponent
{
    constructor(props) {
        super(props);
        this.onChange = this.onChange.bind(this);
    }

    componentDidMount() {
        const { current, onChange } = this.props;
        if (current !== undefined) {
            onChange(this.props.current);
        }
    }

    onChange(current) {
        const { onChange, setState } = this.props;
        setState({ current });
        onChange(current);
    }

    render() {
        const { hidden, items, current } = this.props;

        if (hidden) {
            return null;
        }

        return (
            <SingleSelect
                className="small"
                items={items}
                setState={this.onChange}
                selected={current}
                emptyLabel="Please select"
            />
        );
    }
};

SelectPropertySelect.propTypes = {
    type: PropTypes.oneOf(['select']).isRequired,
    uuid: PropTypes.string.isRequired,
    setState: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    current: PropTypes.any,
    hidden: PropTypes.bool,
};

export default ListsToItemsWrapper(
    CharacterEditorWrapper(
        SelectPropertySelect,
        {
            current: true,
        }
    ),
    'items'
);
