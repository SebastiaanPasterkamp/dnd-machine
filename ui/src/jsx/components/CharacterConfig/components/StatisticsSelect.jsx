import React from 'react';
import Reflux from 'reflux';
import {
    forEach,
    isEqual,
    uniqueId,
} from 'lodash/fp';

import actions from '../actions/CharacterEditorActions.jsx';
import ListDataWrapper from '../../../hocs/ListDataWrapper.jsx';
import store from '../stores/CharacterEditorStore.jsx';

import StatsBlock from '../../StatsBlock.jsx';

class StatisticsSelect extends Reflux.Component
{
    constructor(props) {
        super(props);
        this.store = store;
        this.storeKeys = ['character'];
        this._id = uniqueId();
    }

    componentWillUnmount() {
        super.componentWillUnmount.call(this);
        const { statistics } = this.props;

        actions.removeChange(`${ this._id }-bare`);
        actions.removeChange(`${ this._id }-base`);
        actions.removeChange(`${ this._id }-modifiers`);

        forEach(({ code }) => actions.removeChange(
            `${ this._id }-bonus-${ code }`
        ))(statistics);
    }

    onStatisticsChange = ({ bare, base, modifiers }) => {
        const {
            character: {
                statistics = {},
            },
        } = this.state;

        if (!isEqual(bare, statistics.bare)) {
            actions.addChange(
                'statistics.bare',
                bare,
                `${ this._id }-bare`,
                { type: 'dict' },
            );
        }
        if (!isEqual(base, statistics.base)) {
            actions.addChange(
                'statistics.base',
                base,
                `${ this._id }-base`,
                { type: 'dict' },
            );
        }
        if (!isEqual(modifiers, statistics.modifiers)) {
            actions.addChange(
                'statistics.modifiers',
                modifiers,
                `${ this._id }-modifiers`,
                { type: 'dict' },
            );
        }
    }

    onBonusChange = (bonus) => {
        const { statistics } = this.props;

        forEach(({ code }) => {
            if(code in bonus) {
                actions.addChange(
                    `statistics.bonus.${code}`,
                    {
                        removed: [],
                        added: bonus[code],
                    },
                    `${ this._id }-bonus-${ code }`,
                    {
                        type: 'list',
                        multiple: true,
                    },
                );
            } else {
                actions.removeChange(
                    `${ this._id }-bonus-${ code }`
                );
            }
        })(statistics);
    }

    render() {
        const {
            onChange, getCurrent, current, limit,
            ...props
        } = this.props;
        const {
            character: {
                statistics = {},
            }
        } = this.state;

        return (
            <StatsBlock
                editBase={false}
                {...props}
                {...statistics}
                increase={ limit }
                setState={ this.onStatisticsChange }
                bonusChange={ this.onBonusChange }
            />
        );
    }
};

StatisticsSelect.defaultProps = {
    statistics: [],
};

export default ListDataWrapper(
    StatisticsSelect,
    ['statistics'],
    'items'
);
