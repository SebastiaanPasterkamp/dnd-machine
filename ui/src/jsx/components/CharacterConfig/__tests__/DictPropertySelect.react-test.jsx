import React from 'react';
import { mount } from 'enzyme';

import { DictPropertySelect } from '../components/DictPropertySelect';

const props = {
    type: 'dict',
    uuid: 'mocked-uuid-1',
    path: 'some.path',
    dict: {
        type: 'new',
    },
    current: {
        description: 'This shows _%(type)s_ content',
        type: 'old',
    },
};

describe('Component: DictPropertySelect', () => {

    it('should not render anything while hidden', () => {
        const onChange = jest.fn();
        const wrapper = mount(
            <DictPropertySelect
                onChange={onChange}
                {...props}
                hidden={true}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render the updated description', () => {
        const onChange = jest.fn();
        const wrapper = mount(
            <DictPropertySelect
                onChange={onChange}
                {...props}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should emit onChange dispite being hidden', () => {
        const onChange = jest.fn();
        const wrapper = mount(
            <DictPropertySelect
                onChange={onChange}
                {...props}
                hidden={true}
            />
        );

        expect(onChange).toBeCalledWith(props.dict);
    });
});
