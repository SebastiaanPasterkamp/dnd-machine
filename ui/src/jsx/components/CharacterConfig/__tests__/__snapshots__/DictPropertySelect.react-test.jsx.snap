// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: DictPropertySelect should not render anything while hidden 1`] = `
<DictPropertySelect
  current={
    Object {
      "description": "This shows _%(type)s_ content",
      "type": "old",
    }
  }
  dict={
    Object {
      "type": "new",
    }
  }
  hidden={true}
  onChange={[Function]}
  path="some.path"
  type="dict"
  uuid="mocked-uuid-1"
/>
`;

exports[`Component: DictPropertySelect should render the updated description 1`] = `
<DictPropertySelect
  current={
    Object {
      "description": "This shows _%(type)s_ content",
      "type": "old",
    }
  }
  dict={
    Object {
      "type": "new",
    }
  }
  onChange={[Function]}
  path="some.path"
  type="dict"
  uuid="mocked-uuid-1"
>
  <MDReactComponent
    text="This shows _new_ content"
  >
    <span
      key="mdrct-html-0"
    >
      <p
        key="mdrct-p-0"
      >
        This shows 
        <em
          key="mdrct-em-1"
        >
          new
        </em>
         content
      </p>
    </span>
  </MDReactComponent>
</DictPropertySelect>
`;
